maas/client/viscera/boot_resources.py

```diff
@@ -0,0 +1,285 @@
+        :type filetype: `str`
+        :param chunk_size: Size in bytes to upload to MAAS in chunks.
+            (Default is 4 MiB).
+        :type chunk_size: `int`
+        :param progress_callback: Called to inform the current progress of the
+            upload. One argument is passed with the progress as a precentage.
+            If the resource was already complete and no content
+            needed to be uploaded then this callback will never be called.
+        :type progress_callback: Callable
+        :returns: Create boot resource.
+        :rtype: `BootResource`.
+        """
+        if not isinstance(name, str):
+            raise TypeError("name must be a str, not %s" % type(name).__name__)
+        elif '/' not in name:
+            raise ValueError(
+                "name must be in format os/release; missing '/'")
+        if not isinstance(architecture, str):
+            raise TypeError(
+                "architecture must be a str, not %s" % (
+                    type(architecture).__name__))
+        elif '/' not in architecture:
+            raise ValueError(
+                "architecture must be in format arch/subarch; missing '/'")
+        if not isinstance(content, io.IOBase):
+            raise TypeError(
+                "content must extend from io.IOBase; %s does not" % (
+                    type(content).__name__))
+        elif not content.readable():
+            raise ValueError("content must be readable")
+        elif not content.seekable():
+            raise ValueError("content must be seekable")
+        if title is None:
+            title = ""
+        if not isinstance(title, str):
+            raise TypeError(
+                "title must be a str, not %s" % type(title).__name__)
+        if not isinstance(filetype, BootResourceFiletype):
+            raise TypeError(
+                "filetype must be a BootResourceFiletype, not %s" % (
+                    type(filetype).__name__))
+        if not isinstance(chunk_size, int):
+            raise TypeError(
+                "chunk_size must be a int, not %s" % (
+                    type(chunk_size).__name__))
+        elif chunk_size <= 0:
+            raise ValueError(
+                "chunk_size must be greater than 0, not %d" % chunk_size)
+        if (progress_callback is not None and
+                not isinstance(progress_callback, Callable)):
+            raise TypeError(
+                "progress_callback must be a Callable, not %s" % (
+                type(progress_callback).__name__))
+
+        size, sha256 = calc_size_and_sha265(content, chunk_size)
+        resource = cls._object(cls._handler.create(
+            name=name, architecture=architecture, title=title,
+            filetype=filetype.value, size=str(size), sha256=sha256))
+        newest_set = sorted(resource.sets.keys(), reverse=True)[0]
+        resource_set = resource.sets[newest_set]
+        rfile = list(resource_set.files.values())[0]
```

Comment:
On the face of it this looks buggy. My guess is you're assuming there will always be one file. Add a comment to explain why this is a reasonable thing to do, and be more defensive: at least assert that `len(resource_set.files) == 1` so that it's clear what the assumption is.